Fix MSVC warnings in tests

MarcelRaad opened the pull request:
MSVC complains about narrowing conversions and unreferences parameters.
This makes all tests almost level 4 warning clean on MSVC 14.

Flast commented on test/is_palindrome_test.cpp:
```diff
@@ -18,6 +18,7 @@
 #include <iostream>
 #include <list>
 #include <vector>
+#include <iterator>
```
`std::begin/end` is came from C++11, so I think those should be replaced with `boost::begin/end`.

MarcelRaad commented:
Rebased and removed the two commits which are already integrated. Only the warning fixes are left now.

mclow commented:
IIRC, the `unsigned` in `hex` is there to prevent warnings on other compilers.

mclow commented:
In the `testX` checks, I'd rather change `dest` to a `size_t`, since that's what `std::distance` returns.

MarcelRaad commented:
> IIRC, the unsigned in hex is there to prevent warnings on other compilers.

OK, I reverted that one.

> In the testX checks, I'd rather change dest to a size_t, since that's what std::distance returns.

Right, that's better! It's ptrdiff_t rather than size_t, but it also applies to the operator+ used in the other tests. I changed that now.
